File: src/templates/python/minimal/src/__PACKAGE_NAME__/__CLASS_NAME__.py
#############################################################################
# Author: __EMAIL__
#
# Created on __DATE__
# from template '__TEMPLATE_ID__' of Karabo __KARABO_VERSION__
#
# This file is intended to be used together with Karabo:
#
# http://www.karabo.eu
#
# IF YOU REQUIRE ANY LICENSING AND COPYRIGHT TERMS, PLEASE ADD THEM HERE.
# Karabo itself is licensed under the terms of the MPL 2.0 license.
#############################################################################

from karabo.bound import KARABO_CLASSINFO, PythonDevice

from ._version import version as deviceVersion


@KARABO_CLASSINFO("__CLASS_NAME__", deviceVersion)
class __CLASS_NAME__(PythonDevice):

    @staticmethod
    def expectedParameters(expected):
        """ This static method is needed as a part of the factory/configuration
        system.
        @param expected - Will contain the description of the device's expected
        parameters.
        NOTE: parenthesis () are used for allowing to switch off interpreter
        indentation rule.
        """
        (
        )

    def __init__(self, configuration):
        # always call PythonDevice constructor first!
        super().__init__(configuration)
        # Define the first function to be called after the constructor has
        # finished
        self.registerInitialFunction(self.initialization)

        # If you need methods that can be callable from another device or GUI
        # you may register them here:
        # self.KARABO_SLOT(self.myslot1)
        # self.KARABO_SLOT(myslot2)
        # ...
        # It works for both methods (normal case) and standalone functions.
        # See documentation of KARABO_SLOT for more details.

        # Initialize your member variables here...
        # self._privateData = None

    def initialization(self):
        """
        This method will be called after the constructor.
        """
        # Register methods to process pipeline data.
        # Here 'def onData(self, data, meta)' is registered
        # for InputChannel "input" as defined in expectedParameters.
        # self.KARABO_ON_DATA("input", self.onData)

File: src/templates/cpp/minimal/src/__CLASS_NAME__.cc
/*
 * Author: __EMAIL__
 *
 * Created on __DATE__
 * from template '__TEMPLATE_ID__' of Karabo __KARABO_VERSION__
 *
 * This file is intended to be used together with Karabo:
 *
 * http://www.karabo.eu
 *
 * IF YOU REQUIRE ANY LICENSING AND COPYRIGHT TERMS, PLEASE ADD THEM HERE.
 * Karabo itself is licensed under the terms of the MPL 2.0 license.
 */

#include "__CLASS_NAME__.hh"

#include "karabo/data/schema/NodeElement.hh"
#include "karabo/data/schema/OverwriteElement.hh"
#include "karabo/data/schema/SimpleElement.hh"
#include "karabo/data/schema/TableElement.hh"
#include "karabo/data/schema/VectorElement.hh"
#include "karabo/data/types/Exception.hh"
#include "karabo/data/types/State.hh"
#include "karabo/data/types/Units.hh"
#include "karabo/net/EventLoop.hh"
#include "karabo/xms/SlotElement.hh"


namespace karabo {

    KARABO_REGISTER_FOR_CONFIGURATION(karabo::core::BaseDevice, karabo::core::Device, __CLASS_NAME__)

    void __CLASS_NAME__::expectedParameters(karabo::data::Schema& expected) {
        // Define device schema here, i.e. properties, slots and Input-/OutputChannels
    }

    __CLASS_NAME__::__CLASS_NAME__(const karabo::data::Hash& config) : karabo::core::Device(config) {
        // If the device provides slots (remotely callable methods), add them here.
        // If they should be clickable from the GUI, they have to be added to the schema
        // in expectedParameters in addition.
        // KARABO_SLOT(slotFoo); // void slotFoo() should be a member function

        KARABO_INITIAL_FUNCTION(initialize);
    }


    __CLASS_NAME__::~__CLASS_NAME__() {}


    void __CLASS_NAME__::preReconfigure(karabo::data::Hash& incomingReconfiguration) {}


    void __CLASS_NAME__::postReconfigure() {}

    void __CLASS_NAME__::initialize() {
        // For any InputChannel (e.g. "input") defined in expectedParameters, register
        // a data processing function here. Signature of member function onData should be
        // void onData(const karabo::data::Hash& data, const karabo::xms::InputChannel::MetaData& meta)
        // KARABO_ON_DATA("input", onData);

        // TODO: add any initialization required after the device has been
        //       created with its starting config (e.g. connect to another
        //       device, start some background task, ...). Potentially lengthy
        //       operations should be performed in here, not in the device's
        //       constructor.
    }
} // namespace karabo
